vshard/router/init.lua

```diff
@@ -792,6 +792,76 @@ local function router_call(router, bucket_id, opts, ...)
                             ...)
 end
 
+--
+-- Wait until all masters in the list are connected.
+--
+-- @param replicasets List of replicasets.
+-- @param timeout Timeout in seconds.
+--
+-- @return In case of success - remaining timeout.
+--
+-- @return In case of error - nil, err, UUID of the failed replicaset.
+--
+local function wait_connected_to_masters(replicasets, timeout)
+    local master, uuid, conn
+    local ok, res, err, err_uuid
+
+    -- Start connecting to all masters in parallel.
+    local deadline = fiber_clock() + timeout
+    local connections = table_new(0, #replicasets)
+    for _, replicaset in pairs(replicasets) do
+        master, err = replicaset:wait_master(timeout)
+        uuid = replicaset.uuid
+        if not master then
+            err_uuid = uuid
+            goto fail
+        end
+        conn = replicaset:connect_replica(master)
+        connections[uuid] = conn
+        timeout = deadline - fiber_clock()
+    end
+
+    -- Wait until all connections are established.
+    for uuid, conn in pairs(connections) do
+        -- Fast path. Usually everything is connected.
+        if conn:is_connected() then
+            goto continue
+        end
+
+        -- Slow path. Wait until connected.
+        deadline = fiber_clock() + timeout
+        -- Loop against spurious wakeups.
+        repeat
+            -- Netbox uses fiber_cond inside, which throws an irrelevant usage error
+            -- at negative timeout. Need to check the case manually.
+            if timeout < 0 then
+                err = lerror.timeout()
+                err_uuid = uuid
+                goto fail
+            end
+            ok, res = pcall(conn.wait_connected, conn, timeout)
+            if not ok then
+                err = lerror.make(res)
+                err_uuid = uuid
+                goto fail
+            end
+            if not res then
+                err = lerror.timeout()
+                err_uuid = uuid
+                goto fail
+            end
+            timeout = deadline - fiber_clock()
```

Comment:
Please, don't copy-paste the code. Just use `replicaset:wait_connected`.

Reply:
Done